### PR TITLE
BUG: Add a lock to assert_equal and other testing functions

### PR DESCRIPTION
**Pull request by seberg**
This lock prevents unsafe warning filter manipulations during testing
if downstream packages do parallel testing.

Warning filtering is generally not threadsafe in python, this is also
true for `catch_warnings` or `suppress_warinings`. In NumPy 1.12
however, `assert_equal` and the array comparison asserts, use this
to filter out some comparison warnings. Since removing this filter
may also affect downstream projects and skimage (and possibly more)
do parallel manual parallel testing using `assert_equal`, a quick fix
seems to be to lock the less obvious threading trap. Ideally (in
master this is the case), there should simply not be warning filter
logic in the assert functions themself.
While probably not perfectly safe in principle, it is sufficient
in the case of skimage and probably most testing scenarios and the
chance of deadlocks seems very unlikely.

Closes gh-8413

### CODE REVIEW

File: numpy/testing/utils.py [hunk 1]
```diff
@@ -23,6 +23,25 @@
...
+    # may also affect downstream projects and skimage (and possibly more)
+    # do parallel manual parallel testing using `assert_equal`, a quick fix
+    # seems to be to lock the less obvious threading trap. Ideally (in
+    # master this is the case), there should simply not be warning filter
+    # logic in the assert functions themself.
+    # While probably not perfectly safe in principle, it is sufficient
+    # in the case of skimage and probably most testing scenarios and the
+    # chance of deadlocks seems very unlikely.
+    _assert_comparison_lock = Lock()
+except ImportError:
```
Comment (shoyer) [inline numpy/testing/utils.py]: When is `threading.Lock` not available?

Comment (seberg) [inline numpy/testing/utils.py]: I thought there are some systems/python without threading support or only dummy threading or so?

Comment (seberg) [inline numpy/testing/utils.py]: Though, seems like in fft and some other places, we do not guard against it, in docutils there is some guard using `dummy_threading`.

Comment (juliantaylor) [inline numpy/testing/utils.py]: if python is built without threading, dummy_threading is still available

File: numpy/testing/utils.py [hunk 2]
```diff
@@ -395,13 +414,25 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
...
+    if _assert_comparison_lock is None:
+        # Explicitly use __eq__ for comparison, ticket #2552
+        with suppress_warnings() as sup:
+            # TODO: Better handling will to needed when change happens!
+            sup.filter(DeprecationWarning, ".*NAT ==")
+            sup.filter(FutureWarning, ".*NAT ==")
+            if not (desired == actual):
+                raise AssertionError(msg)
+        return
+    with _assert_comparison_lock:
```
Comment (juliantaylor) [inline numpy/testing/utils.py]: you can put two statements into one with, `with _assert_comparison_lock, suppress_warnings() as sup:`
it looks like the lock it makes more sense inside the warning context manager, but if this is just temporary its probably ok.

File: numpy/testing/utils.py (hunk 1 above)
Comment (seberg) [inline numpy/testing/utils.py]: right, that is nicer then the code duplication then.

File: numpy/testing/utils.py (hunk 2 above)
Comment (seberg) [inline numpy/testing/utils.py]: OK, changed those two things.

Comment (seberg) [inline numpy/testing/utils.py]: I suppose the lock fix could also be done in skimage, its a bit tricky. This only band-aids the cases in skimage, might band-aid some other cases, but it is definitely not a real fix....

**Comment by charris** [comment]
@juliantaylor Look good to you?

**Pull request merged**